crates/wasm-encoder/src/component/types.rs

```diff
@@ -0,0 +1,574 @@
+            Self::Global => bytes.push(GLOBAL_DEFINITION_KIND),
+        }
+    }
+}
+
+/// Represents an index for an outer alias.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum OuterAliasIndex {
+    /// The index is a module index.
+    Module(u32),
+    /// The index is a type index.
+    Type(u32),
+}
+
+impl OuterAliasIndex {
+    pub(crate) fn encode(&self, bytes: &mut Vec<u8>) {
+        match self {
+            Self::Module(index) => {
+                bytes.extend(encoders::u32(*index));
+                bytes.push(OUTER_ALIAS_MODULE);
+            }
+            Self::Type(index) => {
+                bytes.extend(encoders::u32(*index));
+                bytes.push(OUTER_ALIAS_TYPE);
+            }
+        }
+    }
+}
+
+/// Represents an alias.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Alias<'a> {
+    /// The alias is to an instance export.
+    InstanceExport {
+        /// The index of the instance being aliased.
+        instance: u32,
+        /// The name of the export.
+        name: &'a str,
+        /// The kind of the definition being aliased.
+        kind: DefinitionKind,
+    },
+    /// The alias is to an index in an outer module.
+    Outer {
+        /// The count of the outer module.
+        ///
+        /// The count starts at 0 which denotes the current module.
+        count: u32,
+        /// The index from the outer module's index space being aliased.
+        index: OuterAliasIndex,
+    },
+}
+
+impl Alias<'_> {
+    pub(crate) fn encode(&self, bytes: &mut Vec<u8>) {
+        match self {
+            Self::InstanceExport {
+                instance,
+                name,
+                kind,
+            } => {
+                bytes.push(ALIAS_TYPE_INSTANCE_EXPORT);
+                bytes.extend(encoders::u32(*instance));
+                bytes.extend(encoders::str(name));
+                kind.encode(bytes);
+            }
+            Self::Outer { count, index } => {
+                bytes.push(ALIAS_TYPE_OUTER);
+                bytes.extend(encoders::u32(*count));
+                index.encode(bytes);
+            }
+        }
+    }
+}
+
+/// Represents a named interface type.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct NamedInterfaceType<'a> {
+    name: &'a str,
+    ty: InterfaceType,
+}
+
+impl NamedInterfaceType<'_> {
+    pub(crate) fn encode(&self, bytes: &mut Vec<u8>) {
+        bytes.extend(encoders::str(self.name));
+        self.ty.encode(bytes);
+    }
+}
+
+/// Represents results of an adapter function.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum AdapterFunctionResults<'a> {
+    /// The adapter function returns zero results.
+    None,
+    /// The adapter function returns one unnamed result.
+    Single(InterfaceType),
+    /// The function returns multiple named results.
+    ///
+    /// The length of the slice must be greater than one.
+    Multiple(&'a [NamedInterfaceType<'a>]),
```

Comment:
I would naively expect that the names here would be part of a name subsection elsewhere (like the wasm `name` section) rather than directly in the type section itself.

Otherwise though encoding-wise it seems a bit odd here that >1 return value require names and are encoded differently (with names) as opposed to 0 or 1 return values.

Reply:
The names are encoded along with the types so that we can effectively encode everything from a `wit` definition, thus we can generate bindings from the types section when needed.

The reason for not putting it in a custom section is that custom sections are generally seen as optional (and potentially strip-able) for correct behavior; the names obviously don't impact runtime behavior, but they would the DX for users consuming the component.

Regarding the encoding of results from adapter functions this way, that's a very recent change (pushed yesterday) and the intent behind it is to potentially support named return values, but I think there might be more to work out here (for example, we can't encode a single return value with a name). I _believe_ the intention for multiple unnamed results would be to return a single tuple? I'll get more clarification on this too.

Reply:
I was personally assuming that we would go the route of "an auxiliary name-like section must be present for good DX" with the assumption that it still technically works otherwise but it won't be nearly as nice. There's other things from the input `*.wit` file like comments I think we may also want to include (which don't easily have places here). Additionally I think that having a separate section for names would improve symmetry with the core wasm format which personally I thought would be a nice goal to have for interface types and components.

Reply:
I'm open to having a name section too.

Luke and I discussed having a custom section for doc strings since they only improve the DX, whereas the names for types and parameters are pretty essential for even a remotely decent DX.

Let's all hop on a call in the new year to talk this design over.